mpv 0.38.0 Cannot be Built on macOS 10.15

### Important Information

- mpv version: 0.38.0
- Platform version: macOS 10.15 w/ Xcode 12.4
- Meson version: 1.4.1

### Reproduction Steps

The IINA project is planning to use the latest version of mpv for the next release. As mpv claims to support macOS 10.15 or later in the current readme, we decide to build libmpv on a 10.15 mac as the dependency of IINA, and bump the system requirement for IINA to 10.15. We tried to build mpv without any custom flags to reproduce the issue:

- Install dependencies
- `meson setup build`
- `meson compile -C build`

### Expected Behavior

mpv should build normally.

### Actual Behavior

mpv failed to compile due to usage of newer macOS APIs. For instance:
https://github.com/mpv-player/mpv/blob/112fa549beb52ab3d75999c8d8a82936e13001c4/osdep/mac/log_helper.swift#L23
`Logger` is only available for macOS 11+.

See the attached `build_error.log` for more details.

### Log File

[build_error.log](https://github.com/user-attachments/files/15598765/build_error.log)
[meson_setup_build.log](https://github.com/user-attachments/files/15598767/meson_setup_build.log)
[meson-log.txt](https://github.com/user-attachments/files/15598768/meson-log.txt)


### I carefully read all instruction and confirm that I did the following:

- [X] I tested with the latest mpv version to validate that the issue is not already fixed.
- [X] I provided all required information including system and mpv version.
- [X] I produced the log file with the exact same set of parameters, and conditions used in "Reproduction Steps".
- [X] I produced the log file while the behaviors described in "Actual Behavior" were actively observed.
- [X] I attached the full, untruncated log file.

i opened a PR #14334 to fix 2 of the issues. though there are still some others.

syntax problem. not sure yet
```
../osdep/mac/event_helper.swift:78:38: error: expected ',' separator
    func subscribe(_ subscriber: any EventSubscriber, event: Event) {
                                     ^
                                    ,
../osdep/mac/event_helper.swift:78:38: error: unnamed parameters must be written with the empty name '_'
    func subscribe(_ subscriber: any EventSubscriber, event: Event) {
                                     ^
                                     _: 
../osdep/mac/event_helper.swift:78:34: error: cannot find type 'any' in scope
    func subscribe(_ subscriber: any EventSubscriber, event: Event) {
```

can't infer type even though the type should be inferable from the function. probably from the problem above.
```
../video/out/cocoa_cb_common.swift:39:54: error: cannot infer contextual base in reference to member 'init'
        AppHub.shared.event?.subscribe(self, event: .init(name: "MPV_EVENT_SHUTDOWN"))
                                                    ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../osdep/mac/remote_command_center.swift:98:40: error: cannot infer contextual base in reference to member 'init'
        event?.subscribe(self, event: .init(name: "duration", format: MPV_FORMAT_DOUBLE))
                                      ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../osdep/mac/remote_command_center.swift:99:40: error: cannot infer contextual base in reference to member 'init'
        event?.subscribe(self, event: .init(name: "time-pos", format: MPV_FORMAT_DOUBLE))
                                      ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../osdep/mac/remote_command_center.swift:100:40: error: cannot infer contextual base in reference to member 'init'
        event?.subscribe(self, event: .init(name: "speed", format: MPV_FORMAT_DOUBLE))
                                      ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../osdep/mac/remote_command_center.swift:101:40: error: cannot infer contextual base in reference to member 'init'
        event?.subscribe(self, event: .init(name: "pause", format: MPV_FORMAT_FLAG))
                                      ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../osdep/mac/remote_command_center.swift:102:40: error: cannot infer contextual base in reference to member 'init'
        event?.subscribe(self, event: .init(name: "media-title", format: MPV_FORMAT_STRING))
                                      ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../osdep/mac/remote_command_center.swift:103:40: error: cannot infer contextual base in reference to member 'init'
        event?.subscribe(self, event: .init(name: "chapter-metadata/title", format: MPV_FORMAT_STRING))
                                      ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../osdep/mac/remote_command_center.swift:104:40: error: cannot infer contextual base in reference to member 'init'
        event?.subscribe(self, event: .init(name: "metadata/by-key/album", format: MPV_FORMAT_STRING))
                                      ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../osdep/mac/remote_command_center.swift:105:40: error: cannot infer contextual base in reference to member 'init'
        event?.subscribe(self, event: .init(name: "metadata/by-key/artist", format: MPV_FORMAT_STRING))
                                      ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../osdep/mac/touch_bar.swift:144:40: error: cannot infer contextual base in reference to member 'init'
        event?.subscribe(self, event: .init(name: "duration", format: MPV_FORMAT_DOUBLE))
                                      ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../osdep/mac/touch_bar.swift:145:40: error: cannot infer contextual base in reference to member 'init'
        event?.subscribe(self, event: .init(name: "time-pos", format: MPV_FORMAT_DOUBLE))
                                      ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../osdep/mac/touch_bar.swift:146:40: error: cannot infer contextual base in reference to member 'init'
        event?.subscribe(self, event: .init(name: "speed", format: MPV_FORMAT_DOUBLE))
                                      ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../osdep/mac/touch_bar.swift:147:40: error: cannot infer contextual base in reference to member 'init'
        event?.subscribe(self, event: .init(name: "pause", format: MPV_FORMAT_FLAG))
                                      ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../osdep/mac/touch_bar.swift:148:40: error: cannot infer contextual base in reference to member 'init'
        event?.subscribe(self, event: .init(name: "MPV_EVENT_END_FILE"))
                                      ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

with macOS 10.15 compat i meant the latest version. since it uses SDK 11.0 this shouldn't be a problem. though in the worst case we need to add a platform check for this too.
```
../video/out/cocoa_cb_common.swift:138:76: error: 'displayP3_PQ' is only available in macOS 10.15.4 or newer
        case MAC_CSP_DISPLAY_P3_PQ: return CGColorSpace(name: CGColorSpace.displayP3_PQ)
```

i wonder what apple fucked up with the definition of those two here. both are available on 11.0, and both are guarded behind a platform check properly.
https://developer.apple.com/documentation/coregraphics/cgcolorspace/3571187-itur_2100_hlg
https://developer.apple.com/documentation/coregraphics/cgcolorspace/3571188-itur_2100_pq
```
../video/out/cocoa_cb_common.swift:152:78: error: extra argument 'name' in call
            case MAC_CSP_BT_2100_HLG: return CGColorSpace(name: CGColorSpace.itur_2100_HLG)
                                                                ~~~~~~~~~~~~~^~~~~~~~~~~~~
../video/out/cocoa_cb_common.swift:152:58: error: missing arguments for parameters 'calibratedGrayWhitePoint', 'blackPoint', 'gamma' in call
            case MAC_CSP_BT_2100_HLG: return CGColorSpace(name: CGColorSpace.itur_2100_HLG)
                                                         ^
../video/out/cocoa_cb_common.swift:152:78: error: type 'CGColorSpace' has no member 'itur_2100_HLG'
            case MAC_CSP_BT_2100_HLG: return CGColorSpace(name: CGColorSpace.itur_2100_HLG)
                                                                ~~~~~~~~~~~~ ^~~~~~~~~~~~~

../video/out/cocoa_cb_common.swift:153:77: error: extra argument 'name' in call
            case MAC_CSP_BT_2100_PQ: return CGColorSpace(name: CGColorSpace.itur_2100_PQ)
                                                               ~~~~~~~~~~~~~^~~~~~~~~~~~
../video/out/cocoa_cb_common.swift:153:57: error: missing arguments for parameters 'calibratedGrayWhitePoint', 'blackPoint', 'gamma' in call
            case MAC_CSP_BT_2100_PQ: return CGColorSpace(name: CGColorSpace.itur_2100_PQ)
                                                        ^
../video/out/cocoa_cb_common.swift:153:77: error: type 'CGColorSpace' has no member 'itur_2100_PQ'
            case MAC_CSP_BT_2100_PQ: return CGColorSpace(name: CGColorSpace.itur_2100_PQ)
                                                               ~~~~~~~~~~~~ ^~~~~~~~~~~~
```

wonder why this needs to be unwrapped of init expects an optional. something fixed in newer swift versions?
```
../osdep/mac/touch_bar.swift:95:25: error: value of optional type 'NSImage?' must be unwrapped to a value of type 'NSImage'
                image: .init(named: NSImage.touchBarPauseTemplateName),
                        ^
../osdep/mac/touch_bar.swift:95:25: note: coalesce using '??' to provide a default when the optional value contains 'nil'
                image: .init(named: NSImage.touchBarPauseTemplateName),
                        ^
                                                                       ?? <#default value#>
../osdep/mac/touch_bar.swift:95:25: note: force-unwrap using '!' to abort execution if the optional value contains 'nil'
                image: .init(named: NSImage.touchBarPauseTemplateName),
                        ^
                                                                      !
../osdep/mac/touch_bar.swift:96:28: error: value of optional type 'NSImage?' must be unwrapped to a value of type 'NSImage'
                imageAlt: .init(named: NSImage.touchBarPlayTemplateName),
                           ^
../osdep/mac/touch_bar.swift:96:28: note: coalesce using '??' to provide a default when the optional value contains 'nil'
                imageAlt: .init(named: NSImage.touchBarPlayTemplateName),
                           ^
                                                                         ?? <#default value#>
../osdep/mac/touch_bar.swift:96:28: note: force-unwrap using '!' to abort execution if the optional value contains 'nil'
                imageAlt: .init(named: NSImage.touchBarPlayTemplateName),
                           ^
                                                                        !
../osdep/mac/touch_bar.swift:102:25: error: value of optional type 'NSImage?' must be unwrapped to a value of type 'NSImage'
                image: .init(named: NSImage.touchBarGoBackTemplateName),
                        ^
../osdep/mac/touch_bar.swift:102:25: note: coalesce using '??' to provide a default when the optional value contains 'nil'
                image: .init(named: NSImage.touchBarGoBackTemplateName),
                        ^
                                                                        ?? <#default value#>
../osdep/mac/touch_bar.swift:102:25: note: force-unwrap using '!' to abort execution if the optional value contains 'nil'
                image: .init(named: NSImage.touchBarGoBackTemplateName),
                        ^
                                                                       !
../osdep/mac/touch_bar.swift:108:25: error: value of optional type 'NSImage?' must be unwrapped to a value of type 'NSImage'
                image: .init(named: NSImage.touchBarGoForwardTemplateName),
                        ^
../osdep/mac/touch_bar.swift:108:25: note: coalesce using '??' to provide a default when the optional value contains 'nil'
                image: .init(named: NSImage.touchBarGoForwardTemplateName),
                        ^
                                                                           ?? <#default value#>
../osdep/mac/touch_bar.swift:108:25: note: force-unwrap using '!' to abort execution if the optional value contains 'nil'
                image: .init(named: NSImage.touchBarGoForwardTemplateName),
                        ^
                                                                          !
../osdep/mac/touch_bar.swift:114:25: error: value of optional type 'NSImage?' must be unwrapped to a value of type 'NSImage'
                image: .init(named: NSImage.touchBarSkipBackTemplateName),
                        ^
../osdep/mac/touch_bar.swift:114:25: note: coalesce using '??' to provide a default when the optional value contains 'nil'
                image: .init(named: NSImage.touchBarSkipBackTemplateName),
                        ^
                                                                          ?? <#default value#>
../osdep/mac/touch_bar.swift:114:25: note: force-unwrap using '!' to abort execution if the optional value contains 'nil'
                image: .init(named: NSImage.touchBarSkipBackTemplateName),
                        ^
                                                                         !
../osdep/mac/touch_bar.swift:120:25: error: value of optional type 'NSImage?' must be unwrapped to a value of type 'NSImage'
                image: .init(named: NSImage.touchBarSkipAheadTemplateName),
                        ^
../osdep/mac/touch_bar.swift:120:25: note: coalesce using '??' to provide a default when the optional value contains 'nil'
                image: .init(named: NSImage.touchBarSkipAheadTemplateName),
                        ^
                                                                           ?? <#default value#>
../osdep/mac/touch_bar.swift:120:25: note: force-unwrap using '!' to abort execution if the optional value contains 'nil'
                image: .init(named: NSImage.touchBarSkipAheadTemplateName),
                        ^
                                                                          !
../osdep/mac/touch_bar.swift:126:25: error: value of optional type 'NSImage?' must be unwrapped to a value of type 'NSImage'
                image: .init(named: NSImage.touchBarAudioInputTemplateName),
                        ^
../osdep/mac/touch_bar.swift:126:25: note: coalesce using '??' to provide a default when the optional value contains 'nil'
                image: .init(named: NSImage.touchBarAudioInputTemplateName),
                        ^
                                                                            ?? <#default value#>
../osdep/mac/touch_bar.swift:126:25: note: force-unwrap using '!' to abort execution if the optional value contains 'nil'
                image: .init(named: NSImage.touchBarAudioInputTemplateName),
                        ^
                                                                           !
../osdep/mac/touch_bar.swift:132:25: error: value of optional type 'NSImage?' must be unwrapped to a value of type 'NSImage'
                image: .init(named: NSImage.touchBarComposeTemplateName),
                        ^
../osdep/mac/touch_bar.swift:132:25: note: coalesce using '??' to provide a default when the optional value contains 'nil'
                image: .init(named: NSImage.touchBarComposeTemplateName),
                        ^
                                                                         ?? <#default value#>
../osdep/mac/touch_bar.swift:132:25: note: force-unwrap using '!' to abort execution if the optional value contains 'nil'
                image: .init(named: NSImage.touchBarComposeTemplateName),
                        ^
                                                                        !
```

--- a/audio/out/ao_avfoundation.m
+++ b/audio/out/ao_avfoundation.m
@@ -254,9 +254,11 @@ static int init(struct ao *ao)
     }
 
     [p->synchronizer addRenderer:p->renderer];
+#if HAVE_MACOS_11_3_FEATURES
     if (@available(tvOS 14.5, iOS 14.5, macOS 11.3, *)) {
         [p->synchronizer setDelaysRateChangeUntilHasSufficientMediaData:NO];
     }
+#endif
 
     if (af_fmt_is_spdif(ao->format)) {
         MP_FATAL(ao, "avfoundation does not support SPDIF\n");
@@ -315,7 +317,11 @@ static int init(struct ao *ao)
 
     p->observer = [[AVObserver alloc] initWithAO:ao];
     NSNotificationCenter *center = [NSNotificationCenter defaultCenter];
-    [center addObserver:p->observer selector:@selector(handleRestartNotification:) name:AVSampleBufferAudioRendererOutputConfigurationDidChangeNotification object:p->renderer];
+#if HAVE_MACOS_12_FEATURES
+    if (@available(tvOS 15.0, iOS 15.0, macOS 12.0, *)) {
+        [center addObserver:p->observer selector:@selector(handleRestartNotification:) name:AVSampleBufferAudioRendererOutputConfigurationDidChangeNotification object:p->renderer];
+    }
+#endif
     [center addObserver:p->observer selector:@selector(handleRestartNotification:) name:AVSampleBufferAudioRendererWasFlushedAutomaticallyNotification object:p->renderer];
 
     return CONTROL_OK;

--- a/meson.build
+++ b/meson.build
@@ -1585,6 +1585,12 @@ macos_11_features = get_option('macos-11-features').require(
 )
 features += {'macos-11-features': macos_11_features.allowed()}
 
+macos_11_3_features = get_option('macos-11-3-features').require(
+    macos_sdk_version.version_compare('>= 11.3'),
+    error_message: 'A macos sdk version >= 11.3 could not be found!',
+)
+features += {'macos-11-3-features': macos_11_3_features.allowed()}
+
 macos_12_features = get_option('macos-12-features').require(
     macos_sdk_version.version_compare('>= 12'),
     error_message: 'A macos sdk version >= 12 could not be found!',

--- a/meson.build
+++ b/meson.build
@@ -1573,6 +1573,12 @@ if features['cocoa'] and features['swift']
                            'video/out/mac/window.swift')
 endif
 
+macos_10_15_4_features = get_option('macos-10-15-4-features').require(
+    macos_sdk_version.version_compare('>= 10.15.4'),
+    error_message: 'A macos sdk version >= 10.15.4 could not be found!',
+)
+features += {'macos-10-15-4-features': macos_10_15_4_features.allowed()}
+
 macos_11_features = get_option('macos-11-features').require(
     macos_sdk_version.version_compare('>= 11'),
     error_message: 'A macos sdk version >= 11 could not be found!',

--- a/meson_options.txt
+++ b/meson_options.txt
@@ -108,6 +108,7 @@ option('vulkan-interop', type: 'feature', value: 'auto', description: 'Vulkan gr
 # macOS features
 option('macos-10-15-4-features', type: 'feature', value: 'auto', description: 'macOS 10.15.4 SDK Features')
 option('macos-11-features', type: 'feature', value: 'auto', description: 'macOS 11 SDK Features')
+option('macos-11-3-features', type: 'feature', value: 'auto', description: 'macOS 11.3 SDK Features')
 option('macos-12-features', type: 'feature', value: 'auto', description: 'macOS 12 SDK Features')
 option('macos-cocoa-cb', type: 'feature', value: 'auto', description: 'macOS libmpv backend')
 option('macos-media-player', type: 'feature', value: 'auto', description: 'macOS Media Player support')

--- a/meson_options.txt
+++ b/meson_options.txt
@@ -106,6 +106,7 @@ option('videotoolbox-pl', type: 'feature', value: 'auto', description: 'Videotoo
 option('vulkan-interop', type: 'feature', value: 'auto', description: 'Vulkan graphics interop')
 
 # macOS features
+option('macos-10-15-4-features', type: 'feature', value: 'auto', description: 'macOS 10.15.4 SDK Features')
 option('macos-11-features', type: 'feature', value: 'auto', description: 'macOS 11 SDK Features')
 option('macos-12-features', type: 'feature', value: 'auto', description: 'macOS 12 SDK Features')
 option('macos-cocoa-cb', type: 'feature', value: 'auto', description: 'macOS libmpv backend')

--- a/osdep/mac/event_helper.swift
+++ b/osdep/mac/event_helper.swift
@@ -75,7 +75,7 @@ class EventHelper {
         mpv_set_wakeup_callback(mpv, wakeup, TypeHelper.bridge(obj: self))
     }
 
-    func subscribe(_ subscriber: any EventSubscriber, event: Event) {
+    func subscribe(_ subscriber: EventSubscriber, event: Event) {
         guard let mpv = mpv else { return }
 
         if !event.name.isEmpty {

--- a/osdep/mac/log_helper.swift
+++ b/osdep/mac/log_helper.swift
@@ -20,7 +20,12 @@ import os
 
 class LogHelper {
     var log: OpaquePointer?
-    let logger = Logger(subsystem: "io.mpv", category: "mpv")
+#if HAVE_MACOS_11_FEATURES
+    @available(macOS 11.0, *)
+    var logger: Logger? {
+        return Logger(subsystem: "io.mpv", category: "mpv")
+    }
+#endif
 
     let loggerMapping: [Int: OSLogType] = [
         MSGL_V: .debug,
@@ -51,7 +56,11 @@ class LogHelper {
 
     func send(message: String, type: Int) {
         guard let log = log else {
-            logger.log(level: loggerMapping[type] ?? .default, "\(message, privacy: .public)")
+#if HAVE_MACOS_11_FEATURES
+        if #available(macOS 11.0, *) {
+            logger?.log(level: loggerMapping[type] ?? .default, "\(message, privacy: .public)")
+        }
+#endif
             return
         }
 

--- a/osdep/mac/meson.build
+++ b/osdep/mac/meson.build
@@ -19,6 +19,10 @@ if get_option('optimization') != '0'
     swift_flags += '-O'
 endif
 
+if macos_10_15_4_features.allowed()
+    swift_flags += ['-D', 'HAVE_MACOS_10_15_4_FEATURES']
+endif
+
 if macos_11_features.allowed()
     swift_flags += ['-D', 'HAVE_MACOS_11_FEATURES']
 endif

--- a/osdep/mac/swift_compat.swift
+++ b/osdep/mac/swift_compat.swift
@@ -43,3 +43,11 @@ extension NSDraggingInfo {
     }
 }
 #endif
+
+#if !HAVE_MACOS_12_FEATURES && HAVE_MACOS_11_FEATURES
+@available(macOS 11.0, *)
+extension CGColorSpace {
+    static let itur_2100_HLG: CFString = kCGColorSpaceITUR_2100_HLG
+    static let itur_2100_PQ: CFString = kCGColorSpaceITUR_2100_PQ
+}
+#endif

--- a/osdep/mac/swift_compat.swift
+++ b/osdep/mac/swift_compat.swift
@@ -15,6 +15,16 @@
  * License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+#if !swift(>=5.7)
+extension NSCondition {
+    func withLock<R>(_ body: () throws -> R) rethrows -> R {
+        self.lock()
+        defer { self.unlock() }
+        return try body()
+    }
+}
+#endif
+
 #if !swift(>=5.0)
 extension Data {
     mutating func withUnsafeMutableBytes<Type>(_ body: (UnsafeMutableRawBufferPointer) throws -> Type) rethrows -> Type {

--- a/video/out/cocoa_cb_common.swift
+++ b/video/out/cocoa_cb_common.swift
@@ -135,7 +135,6 @@ class CocoaCB: Common, EventSubscriber {
         case MAC_CSP_AUTO: return colorSpace
         case MAC_CSP_DISPLAY_P3: return CGColorSpace(name: CGColorSpace.displayP3)
         case MAC_CSP_DISPLAY_P3_HLG: return CGColorSpace(name: CGColorSpace.displayP3_HLG)
-        case MAC_CSP_DISPLAY_P3_PQ: return CGColorSpace(name: CGColorSpace.displayP3_PQ)
         case MAC_CSP_DCI_P3: return CGColorSpace(name: CGColorSpace.dcip3)
         case MAC_CSP_BT_2020: return CGColorSpace(name: CGColorSpace.itur_2020)
         case MAC_CSP_BT_709: return CGColorSpace(name: CGColorSpace.itur_709)
@@ -145,6 +144,15 @@ class CocoaCB: Common, EventSubscriber {
         case MAC_CSP_ADOBE: return CGColorSpace(name: CGColorSpace.adobeRGB1998)
         default: break
         }
+
+#if HAVE_MACOS_10_15_4_FEATURES
+        if #available(macOS 10.15.4, *) {
+            switch outputCsp {
+            case MAC_CSP_DISPLAY_P3_PQ: return CGColorSpace(name: CGColorSpace.displayP3_PQ)
+            default: break
+            }
+        }
+#endif
 
 #if HAVE_MACOS_11_FEATURES
         if #available(macOS 11.0, *) {

